ci: Support running tests using ARM FVP emulator

stephanosio opened the pull request:
This series adds support for running tests using the ARM FVP emulator in the CI.

The ARM FVP cannot be included in the public CI Docker image due to licensing issues (redistribution is prohibited by the EULA), so a private CI Docker image (`docker-internal`) has been created.

The private CI Docker image includes the following ARM FVP models (note that the scope of this image may be expanded to include other proprietary toolchains in the future):

* FVP_BaseR_AEMv8R
* FVP_Base_RevC-2xAEMvA
* FVP_Corstone_SSE-300_Ethos-U55

With this, the following platforms, some of which were not being run at all, can now be tested in the CI:

* fvp_baser_aemv8r
* fvp_base_revc_2xaemv8a
* mps3_an547 (was previously using QEMU, but now uses FVP with more advanced emulation features such as the emulation of the Ethos-U NPU).

Closes #43852

Requires #45108, #45110, #45111 to be fixed prior to merging

stephanosio committed:
boards: fvp_baser_aemv8r: Enable simulation with ARM FVP

This commit adds `simulation: armfvp` in the board YAML files so that
twister runs the tests for these boards using the ARM FVP emulator.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
boards: fvp_base_revc_2xaemv8a: Enable simulation with ARM FVP

This commit adds `simulation: armfvp` in the board YAML file so that
twister runs the tests for this board using the ARM FVP emulator.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
boards: fvp_baser_aemv8r_aarch32: Enable simulation with ARM FVP

This commit adds `simulation: armfvp` in the board YAML file so that
twister runs the tests for this board using the ARM FVP emulator.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
boards: mps3_an547: Change default emulator to ARM FVP

This commit changes the default emulator for the `mps3_an547` boards
from QEMU to ARM FVP, which is more accurate and supports additional
features (e.g. Ethos NPU emulation).

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
boards: mps3_an547: Enable as default testing platform

This commit enables the `mps3_an547` board as one of the default
testing platforms because it is a reference platform for the Cortex-M55
and Ethos-U cores, and there are no other default testing platforms
that support these cores.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
boards: fvp_baser_aemv8r_aarch32: Enable as default testing platform

This commit enables the `fvp_baser_aemv8r_aarch32` board as one of the
default testing platforms because it is a reference platform for the
Armv8-R cores running in the AArch32 mode, and there are no other
default testing platforms that cover this configuration.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
boards: fvp_baser_aemv8r: Enable as default testing platform

This commit enables the `fvp_baser_aemv8r` and `fvp_baser_aemv8r_smp`
boards as default testing platforms because it is a reference platform
for the Armv8-R cores running in the AArch64 mode, and there are no
other default testing platforms that cover this configuration.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
boards: fvp_base_revc_2xaemv8a: Enable as default testing platform

This commit enables the `fvp_base_revc_2xaemv8a` board as one of the
default testing platforms because it is a reference platform for the
Armv8-A cores running in the AArch64 mode, and there are no other
default testing platforms that cover this configuration.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
ci: Switch to private CI image

This commit changes the CI workflows to use the private (internal) CI
docker image v0.23.3.0, which is based on the public CI docker image
v0.23.3.

The private CI image contains additional tools that cannot be included
in the public CI image due to licensing issues.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

povergoing commented on boards/arm64/fvp_baser_aemv8r/fvp_baser_aemv8r.yaml:
```diff
@@ -5,8 +5,11 @@ identifier: fvp_baser_aemv8r
 name: FVP Emulation FVP_BaseR_AEMv8R
 arch: arm64
 type: sim
+simulation: armfvp
```
Is that our fixes should be merged right before this PR? Or I am afraid this would block others?

And I think we need to add a warning or alerting when users run `twister `locally  without installing any FVP... I'll take a look

stephanosio replied:
> Is that our fixes should be merged right before this PR? Or I am afraid this would block others?

Yes, the fixes need to be merged first -- https://github.com/zephyrproject-rtos/zephyr/issues/43852#issuecomment-1108231033.

> And I think we need to add a warning or alerting when users run twister locally without installing any FVP.

Indeed, we should.

stephanosio replied:
> And I think we need to add a warning or alerting when users run twister locally without installing any FVP.

Added https://github.com/zephyrproject-rtos/zephyr/pull/45099/commits/8683cac21d2fbdd73d721c8f9ee54dbfdcbb0e9b, which makes twister skip the platforms that require the ARM FVP when the `ARMFVP_BIN_PATH` environment variable is not set.

stephanosio commented:
> so, if this PR gets merged, it means that tests will run of FVP as well?

@carlocaione Yes, CI will be testing the platforms listed in https://github.com/zephyrproject-rtos/zephyr/pull/45099#issue-1213688302 using the ARM FVP.

Note that there are still some failures that need to be fixed prior to merging this:
https://github.com/zephyrproject-rtos/zephyr-testing/actions/runs/2216000834

microbuilder commented:
> The ARM FVP cannot be included in the public CI Docker image due to licensing issues (redistribution is prohibited by the EULA), so a private CI Docker image (`docker-internal`) has been created.

I believe the Corstone images (at least the AN547, M55+U55 one) can actually be used in public CI and have a different license from the large FVP collection, which require a license to be purchased:  https://developer.arm.com/tools-and-software/open-source-software/arm-platforms-software/arm-ecosystem-fvps

```
Ecosystem FVPs are available without license control for direct download from the following table.
They are supported by relevant Open Source Software projects. 
```

They can even be downloaded directly in a way that is likely CI-friendly.

microbuilder commented:
Somewhat of a sidenote, but the Corstone-1000 FVP looks like it might be useful for testing if we're enabling support for emulation since it's a Cortex-M0+ and a Cortex A35, where the device starts up on the M0+ which then starts the A35.

It could be an interesting testbed for Cortex-A and something like OpenAMP for A+M communication in CI.

stephanosio commented:
> I believe the Corstone (Ecosystem) images (at least the AN547, M55+U55 one) can actually be used in public CI and have a different license from the large FVP collection

@microbuilder The M55+U55/U65 one (SSE-300) is already used by `mps3_an547` in this PR.

The problem was that the EULA, even for free versions, does not allow redistribution: https://github.com/zephyrproject-rtos/zephyr/issues/43852#issuecomment-1068928542. With the new approach using a private CI Docker image, we do not redistribute the FVP.

> Somewhat of a sidenote, but the Corstone-1000 FVP looks like it might be useful for testing if we're enabling support for emulation since it's a Cortex-M0+ and a Cortex A35, where the device starts up on the M0+ which then starts the A35.

That would be very useful indeed. If such a board support is added to the Zephyr in the future, we can look into including the Corstone-1000 FVP in the private CI image as well.

stephanosio committed:
twister: Skip running 'armfvp' tests if ARMFVP_BIN_PATH is not set

This commit updates the twisterlib to skip running the tests for the
platforms that require the ARM FVP emulator if the `ARMFVP_BIN_PATH`
environment variable is not set.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

povergoing commented on boards/arm64/fvp_baser_aemv8r/fvp_baser_aemv8r_smp.yaml:
```diff
@@ -5,8 +5,11 @@ identifier: fvp_baser_aemv8r_smp
 name: FVP Emulation FVP_BaseR_AEMv8R
 arch: arm64
 type: sim
+simulation: armfvp
```
@stephanosio  Can we remove this one, and skip `fvp_baser_aemv8r_smp` at this time. I may need more time to fix the SMP board. I can add it back when I fix all the test cases for `fvp_baser_aemv8r_smp`.

stephanosio replied:
Since this is not urgent, it can wait until `fvp_baser_aemv8r_smp` is fixed.

stephanosio commented:
Rebased

stephanosio commented:
@povergoing I am seeing a failure after rebase:

```
INFO    - 1160/1160 fvp_base_revc_2xaemv8a    tests/kernel/usage/thread_runtime_stats/kernel.usage FAILED Failed (armfvp 3.568s)
...
START - test_thread_stats_usage
    Assertion failed at WEST_TOPDIR/zephyr-testing/tests/kernel/usage/thread_runtime_stats/src/test_thread_runtime_stats.c:517: test_thread_stats_usage: stats5.current_cycles < stats3.current_cycles is false
 FAIL - test_thread_stats_usage in 0.57 seconds
```

https://github.com/zephyrproject-rtos/zephyr-testing/runs/6745649272?check_suite_focus=true#step:11:1593

povergoing commented:
> @povergoing I am seeing a failure after rebase:
> 
> ```
> INFO    - 1160/1160 fvp_base_revc_2xaemv8a    tests/kernel/usage/thread_runtime_stats/kernel.usage FAILED Failed (armfvp 3.568s)
> ...
> START - test_thread_stats_usage
>     Assertion failed at WEST_TOPDIR/zephyr-testing/tests/kernel/usage/thread_runtime_stats/src/test_thread_runtime_stats.c:517: test_thread_stats_usage: stats5.current_cycles < stats3.current_cycles is false
>  FAIL - test_thread_stats_usage in 0.57 seconds
> ```
> 
> https://github.com/zephyrproject-rtos/zephyr-testing/runs/6745649272?check_suite_focus=true#step:11:1593
> 
> Note that this is a newly added test; see #45489
> 
> cc @peter-mitsis
> 
> p.s. This also fails on `fvp_baser_aemv8r`; interestingly, it does not fail on `fvp_baser_aemv8r_aarch32`.

I'll take a look

stephanosio commented:
Rebased

stephanosio commented:
Tested in https://github.com/zephyrproject-rtos/zephyr-testing/actions/runs/2587627407

stephanosio commented:
At last, all tests passed in the CI: https://github.com/zephyrproject-rtos/zephyr-testing/actions/runs/2587627407.

@povergoing @SgrrZhf @carlocaione @microbuilder please review

microbuilder reviewed (dismissed):
Thanks for getting this in!

povergoing reviewed (dismissed):
Thumbs up. Thanks so much!

carlescufi reviewed (dismissed):
These workflows are being reused downstream by others. Is it possible to limit the use of the internal image to those workflows that need it?

PerMac commented:
> > These workflows are being reused downstream by others. Is it possible to limit the use of the internal image to those workflows that need it?
> 
> sure, I have reworked the patch such that the internal image is only used by the `clang.yaml`, `codecov.yaml` and `twister.yaml` workflows which actually require it. Note that these workflows require `zephyr_runner` so they cannot be used downstream as-is anyway.

We already have a patch, switching the node back to github defaults and adding parallel execution. Why those workflows require the internal version now? Is it due to adding `testing:
  default: true` to these simulators?

stephanosio commented:
@carlescufi could you revisit this?

carlescufi commented:
> @carlescufi could you revisit this?

sorry, do you mind rebasing @stephanosio ?

nashif commented on .github/workflows/codecov.yaml:
```diff
@@ -18,7 +18,7 @@ jobs:
     needs: codecov-prep
     container:
-      image: ghcr.io/zephyrproject-rtos/ci:v0.23.3
+      image: ghcr.io/zephyrproject-rtos/ci-internal:v0.23.3.0
```
We should not use internal/private  docker images for the common and major CI workflows. We need to a better understanding of why we are doing this and what is the impact on users trying to reproduce an issue that happend in CI without having to go install 3rd part stuff that is not part of the documented environemt or the public docker images available for everyone. Marking this to be discussed in the TSC this week.

stephanosio replied:
> what is the impact on users trying to reproduce an issue that happend in CI

The "private/internal CI docker image" is what contains the tools (in this case, the ARM FVP emulator) that cannot be included in the public CI image due to [the licensing issues](https://github.com/zephyrproject-rtos/zephyr/issues/43852#issuecomment-1068928542) (redistribution is not allowed).

The ARM FVP emulator included in this image is publicly available and free for everyone to download (only redistribution is not allowed), so anyone can go on the ARM website, download the ARM FVP, and reproduce any issues seen in the CI if ever needed; moreover, they have now changed the download process such that registration is no longer necessary.

> that is not part of the documented environemt

The instructions for downloading the ARM FVP is [documented for the boards that require the ARM FVP](https://github.com/zephyrproject-rtos/zephyr/blame/30cd0d4338601ffd39d8f5d1a992cb4f090ced6e/boards/arm/fvp_baser_aemv8r_aarch32/doc/index.rst#L17-L19).

> We need to a better understanding of why we are doing this 

There is a great value in including the ARM FVP in the CI because various ARM architecture variants and reference platforms require this tool for emulation (and, subsequently, for testing); for example, "ARMv8-R architecture port" is completely untested in the CI without the ARM FVP and we have seen various breakages in the upstream code because of this.

> Marking this to be discussed in the TSC this week.

In essence, we need to decide between:

a) having the upstream ARM architecture ports and reference platforms well tested in the CI at the cost of the minor inconvenience for a small number of users to manually download and install the ARM FVP if they ever need it for local testing

b) not testing the upstream ARM architecture ports and reference platforms in the CI and letting them rot away as we have previously seen

Given the importance of the ARM architecture support to many stakeholders and the project itself, I think the importance of this does not need to be reiterated.

Moreover, the private/internal CI docker image would eventually be something inevitable as we start adding more third-party tooling support anyway (e.g. other third-party emulators and toolchains), but let us leave that discussion for another day.

p.s. It would certainly require further discussion for the private tools that requires registration and approval like, for instance, the nSIM (for ARC); but, in the case of the ARM FVP, it is both public and freely available.

nashif commented:
the issue is have is not what we are testing and the value we get from testing this, it is basically this:

>      image: ghcr.io/zephyrproject-rtos/ci:v0.23.3
>     image: ghcr.io/zephyrproject-rtos/ci-internal:v0.23.3.0

This makes our docker images private, which have been open and public for years, this is a 'regression' IMO and can be confusing. Also, not that is it private, the tempatation to add more closed stuff is just going to grow. As I said, I am all of testing, but at not at any cost and not without having this gone a thorrogh reviews and discussed alteranives.

stephanosio commented:
> This makes our docker images private, which have been open and public for years

The [internal CI Docker image](https://github.com/zephyrproject-rtos/docker-ci-internal/blob/cc116c53d7c50fd32143ed82d534b397171d9593/Dockerfile#L1) is based off the [public CI Docker image](https://github.com/zephyrproject-rtos/docker-image) -- it basically takes the public image and installs the ARM FVP on top of it.

In the proposed scheme, the public CI Docker image will always be kept up to date and the downstream Zephyr users will be able to keep using the public CI Docker image in their CI (the tests for the platforms that require the ARM FVP will be automatically skipped).

Also note that the [`Dockerfile` for the internal CI Docker image](https://github.com/zephyrproject-rtos/docker-ci-internal/blob/master/Dockerfile) is public, so anyone can clone the repository and build the Docker image for local testing or for use in their own CI as long as they do not redistribute it in the binary form.

nashif commented:
I do not understand why https://github.com/zephyrproject-rtos/docker-ci-internal/blob/cc116c53d7c50fd32143ed82d534b397171d9593/Dockerfile#L1 is private then.  Why not make this public, as a Docker file and make it available, you can also allow developers to just build it themselves, that should not vioiate the redistribution rules, the only thing that would not work is someone pulling the docker images with ARM FVP, which is fine, but everything else could be open.

nashif commented:
> Also note that the [`Dockerfile` for the internal CI Docker image](https://github.com/zephyrproject-rtos/docker-ci-internal/blob/master/Dockerfile?rgh-link-date=2022-07-11T14%3A54%3A28Z) is public, so anyone can clone the repository and build the Docker image for local testing or for use in their own CI as long as they do not redistribute it in the binary form.

I should have read to the end first, but this good, we just end up having too many repos, why not put it next to the other Dockerfile?

nashif commented:
> 1. the issues with the "internal" image can be tracked more nicely in a dedicated repository. also the people who are only interested in the public image do not get notified about the issues in the internal image this way.

The point I am trying to make is that everything is public less the built docker image itself, i.e. you can't get full access to the built image, but you can build it yourself and you can do whatever you want with it. So it less confusing and more transparent to have the source maintained and managed in a consitent way, in one place if possible, to avoid the possible misconception that we are doing something that is 'internal' only.

stephanosio commented:
As per the discussion in the TSC meeting on 21 July 2022, this PR will be reworked to introduce a separate CI workflow using the "internal" image with a broader scheme to support the proprietary tooling including, but not limited to, the ARM FVP.

mmahadevan108 commented:
@stephanosio I have cleared the 3.2 milestone as we are past the feature freeze date. Kindly pick a milestone for a future release.

povergoing commented:
> As per the discussion in the TSC meeting on 21 July 2022, this PR will be reworked to introduce a separate CI workflow using the "internal" image with a broader scheme to support the proprietary tooling including, but not limited to, the ARM FVP.

@stephanosio May I ask what the separate CI workflow will look like? Is there anything different with your first proposal, the "private/internal CI docker image"?

stephanosio commented:
> @stephanosio May I ask what the separate CI workflow will look like? Is there anything different with your first proposal, the "private/internal CI docker image"?

It should look about the same as what we have in this PR right now -- it will just be in a separate dedicated workflow instead of the main `twister.yaml` workflow.

However, there are some logistical problems with the separate workflow approach, which we will need to solve before we get to implementing it:

* minimising the amount of duplicate code among the main twister workflow (`twister.yaml`) and the twister workflow(s) using proprietary tools.
* properly scoping the tests that are built and run in a workflow so as to eliminate or minimise the amount of duplicate test builds and runs across multiple workflows.

carlocaione commented:
@stephanosio what should we do with this?

stephanosio commented:
> @stephanosio what should we do with this?

@carlocaione The PR, in its current form, has been rejected by the TSC. I would let it close since this will need to be completely reworked and there is a dedicated issue tracking the enhancement (https://github.com/zephyrproject-rtos/zephyr/issues/43852).

microbuilder commented:
Closing rather than just letting it die a slow death by stale bot, kept alive by comments. 😉
